geaflow/geaflow-dsl/geaflow-dsl-plan/src/main/java/com/antgroup/geaflow/dsl/udf/graph/TriangleCount.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.antgroup.geaflow.dsl.udf.graph;
+
+import com.antgroup.geaflow.common.type.primitive.LongType;
+import com.antgroup.geaflow.dsl.common.algo.AlgorithmRuntimeContext;
+import com.antgroup.geaflow.dsl.common.algo.AlgorithmUserFunction;
+import com.antgroup.geaflow.dsl.common.data.Row;
+import com.antgroup.geaflow.dsl.common.data.RowEdge;
+import com.antgroup.geaflow.dsl.common.data.RowVertex;
+import com.antgroup.geaflow.dsl.common.data.impl.ObjectRow;
+import com.antgroup.geaflow.dsl.common.function.Description;
+import com.antgroup.geaflow.dsl.common.types.StructType;
+import com.antgroup.geaflow.dsl.common.types.TableField;
+import com.antgroup.geaflow.model.graph.edge.EdgeDirection;
+import com.antgroup.geaflow.model.graph.edge.IEdge;
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Description(name = "triangle_count", description = "built-in udga for Triangle Count.")
+public class TriangleCount implements AlgorithmUserFunction<Object, ObjectRow> {
+
+    private AlgorithmRuntimeContext<Object, ObjectRow> context;
+
+    private final int maxIteration = 2;
+
+    private String vertexType = null;
+
+    @Override
+    public void init(AlgorithmRuntimeContext<Object, ObjectRow> context, Object[] params) {
+        this.context = context;
+        if (params.length >= 1) {
+            assert params[0] instanceof String : "Vertex type parameter should be string.";
+            vertexType = (String) params[0];
+        }
+    }
+
+    @Override
+    public void process(RowVertex vertex, Iterator<ObjectRow> messages) {
+        if (Objects.nonNull(vertexType) && !vertexType.equals(vertex.getLabel())) {
+            return;
+        }
+
+        if (context.getCurrentIterationId() == 1L) {
+            List<RowEdge> rowEdges = context.loadEdges(EdgeDirection.BOTH);
+            List<Object> targetIds = rowEdges.stream().map(IEdge::getTargetId).collect(Collectors.toList());
+            targetIds.add(0, (long) targetIds.size());
+            ObjectRow msg = ObjectRow.create(targetIds.toArray());
+            for (int i = 1; i < targetIds.size(); i++) {
+                context.sendMessage(targetIds.get(i), msg);
+            }
+            context.updateVertexValue(msg);
+        } else if (context.getCurrentIterationId() <= maxIteration) {
+            long count = 0;
+            Set<Long> sourceSet = row2Set(vertex.getValue());
+            while (messages.hasNext()) {
+                ObjectRow msg = messages.next();
+                Set<Long> targetSet = row2Set(msg);
+                targetSet.retainAll(sourceSet);
```

Comment:
There is no restriction on the type of the third  here, which may seem a bit strange when specifying type parameters. The intersection of targetIds of two vertices of type A may occur on a vertex of type B, which makes the triangle counting not only happen on vertices of type A, which is different from the general understanding.

Reply:
Thanks, yeah, if specify `vertex type`, it does happen, should I remove the `vertex_type` parameter now or increase the `edge_type` parameter to limit the graph structure because can't get `vertex_label` by `vertex_id`.

Reply:
> Thanks, yeah, if specify `vertex type`, it does happen, should I remove the `vertex_type` parameter now or increase the `edge_type` parameter to limit the graph structure because can't get `vertex_label` by `vertex_id`.

Vertex that do not meet the type requirements will not send messages to their neighbors. Perhaps this can be used to exclude the vertex that do not belong to the required type by receiving messages and creating a set.

Reply:
Hi, @Leomrlin, could you help me review the code in your free time again, thanks!

Reply:
> Hi, @Leomrlin, could you help me review the code in your free time again, thanks!
Of course, however, using excludeSet is not a good solution. Please see my suggestion for details